xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp

```diff
@@ -1163,8 +1163,14 @@ bool CDVDDemuxFFmpeg::SeekTime(double time, bool backwards, double* startpts)
   int64_t seek_pts = (int64_t)time * (AV_TIME_BASE / 1000);
   bool ismp3 = m_pFormatContext->iformat && (strcmp(m_pFormatContext->iformat->name, "mp3") == 0);
 
-  if (m_checkvideo)
+  CLog::Log(LOGNOTICE, "%s - Do I need to check the video: %d, and which stream do I seek: %d", __FUNCTION__, m_checkvideo, m_seekStream);
+
+  if (m_checkvideo)// && IsVideoReady())
```

Comment:
if IsVideoReady is false, read from demuxer until it becomes  true. they required parameters are set

Reply:
Ah, it’s that easy. Thanks

Reply:
Was hoping it might be as simple as:

```
  if (m_checkvideo))
  {
    if (!IsVideoReady())
    {
      CLog::Log(LOGNOTICE, "%s - Video not ready, reading", __FUNCTION__);
      Read();
    }

    AVStream* st = m_pFormatContext->streams[m_seekStream];
    seek_pts = av_rescale(m_startTime + time / 1000, st->time_base.den, st->time_base.num);
  }
```

Reply:
if -> while
further more you should evaluate the return value of Read. in case of empty packet sleep 10ms. in case of IsVideoready does not become true after a while (maybe 1s), exit the loop and make Seek return failure

Reply:
don't forget to free the returned demux packet: CDVDDemuxUtils::FreeDemuxPacket

Reply:
Thanks again 😉

Reply:
It was a while but even failing with an if.

Reply:
```
--- a/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp
+++ b/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp
@@ -31,6 +31,10 @@
 #include "utils/StringUtils.h"
 #include "utils/URIUtils.h"
 
+#ifdef TARGET_POSIX
+#include "platform/posix/XTimeUtils.h"
+#endif
+
 #ifdef HAVE_LIBBLURAY
 #include "DVDInputStreams/DVDInputStreamBluray.h"
 #endif
@@ -1165,6 +1169,16 @@ bool CDVDDemuxFFmpeg::SeekTime(double time, bool backwards, double* startpts)
 
   if (m_checkvideo)
   {
+    while (!IsVideoReady())
+    {
+      DemuxPacket *pkt = Read();
+      if (pkt)
+        CDVDDemuxUtils::FreeDemuxPacket(pkt);
+      else
+        Sleep(10);
+      m_pkt.result = -1;
+      av_packet_unref(&m_pkt.pkt);
+    }
     AVStream* st = m_pFormatContext->streams[m_seekStream];
     seek_pts = av_rescale(m_startTime + time / 1000, st->time_base.den, st->time_base.num);
   }
@@ -2135,7 +2149,7 @@ bool CDVDDemuxFFmpeg::IsVideoReady()
         {
           if (!m_startTime)
           {
-            m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den);
+            m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
             m_seekStream = i;
           }
           return true;
@@ -2153,7 +2167,7 @@ bool CDVDDemuxFFmpeg::IsVideoReady()
         st = m_pFormatContext->streams[idx];
         if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO)
         {
-          m_startTime = static_cast<double>(av_rescale(st->cur_dts, st->time_base.num, st->time_base.den));
+          m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
           m_seekStream = i;
           break;
         }
@@ -2171,7 +2185,7 @@ bool CDVDDemuxFFmpeg::IsVideoReady()
         {
           if (!m_startTime)
           {
-            m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den);
+            m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
             m_seekStream = i;
           }
           return true;
@@ -2188,7 +2202,7 @@ bool CDVDDemuxFFmpeg::IsVideoReady()
         st = m_pFormatContext->streams[i];
         if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO)
         {
-          m_startTime = static_cast<double>(av_rescale(st->cur_dts, st->time_base.num, st->time_base.den));
+          m_startTime = av_rescale(st->cur_dts, st->time_base.num, st->time_base.den) - 0.000001;
           m_seekStream = i;
           break;
         }

```

Reply:
Ok, not so far off my version.

As I'm learning can you explain this:
```
      m_pkt.result = -1;
```
      
I was trying to call FreeDemuxPacket which was incorrect I guess instead of `av_packet_unref(&m_pkt.pkt)`;

Secondly why the subtraction of  0.000001 in IsVideoReady()?

My version also has the 1 second check so will combine and push shortly.

Reply:
ffmpeg's mpeg-ts demuxer is a bit crappy. mpeg-ts signals a stream change before reading next packet but with ffmpeg you have to read a packet in oder to detect a stream change. this is why CDVDDemuxUtils buffers the last packet on a stream change. setting result to -1 and av_packet_unref releases the internal buffer. if it is not released, Read() get this last packet.

> Lastly why the subtraction of 0.000001 in IsVideoReady()?

because m_startTime is compared to 0 and an actual start time of 0 results in permanently overwriting m_startTime

Reply:
Thanks, makes sense.

Is my 1 second check sufficient?